### ISSUE TITLE
Stock Prediction

### ISSUE THREAD

**Issue by CHEVURIVENKATESH2709** [CHEVURIVENKATESH2709/CHEVURIVENKATESH#1]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import yfinance as yf
from sklearn.preprocessing import MinMaxScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout

# Set the ticker symbol of the company you want to predict (e.g., 'AAPL' for Apple Inc.)
ticker_symbol = 'AAPL'

# Define a function to fetch stock price data from Yahoo Finance
def fetch_stock_data(ticker, start_date, end_date):
    data = yf.download(ticker, start=start_date, end=end_date)
    return data

# Fetch stock price data for a specific time range
start_date = '2020-01-01'
end_date = '2023-01-01'
stock_data = fetch_stock_data(ticker_symbol, start_date, end_date)

# Plot the stock price data
plt.figure(figsize=(12, 6))
plt.title(f'{ticker_symbol} Stock Price')
plt.xlabel('Date')
plt.ylabel('Price (USD)')
plt.plot(stock_data['Close'])
plt.show()

# Preprocess the data
scaler = MinMaxScaler()
scaled_data = scaler.fit_transform(stock_data['Close'].values.reshape(-1, 1))

training_data_len = int(np.ceil( len(scaled_data) * .95 ))

train_data = scaled_data[0:int(training_data_len), :]

x_train=[]
y_train = []
for i in range(60,len(train_data)):
    x_train.append(train_data[i-60:i,0])
    y_train.append(train_data[i,0])
    
x_train, y_train = np.array(x_train), np.array(y_train)

x_train = np.reshape(x_train, (x_train.shape[0],x_train.shape[1],1))

# Build the LSTM model
model = Sequential()

model.add(LSTM(units=50, return_sequences=True, input_shape=(x_train.shape[1],1)))
model.add(LSTM(units=50, return_sequences=False))
model.add(Dense(units=25))
model.add(Dense(units=1))

# Compile the model
model.compile(optimizer='adam', loss='mean_squared_error')

# Train the model
model.fit(x_train, y_train, batch_size=1, epochs=1)

# Test data set
test_data = scaled_data[training_data_len - 60: , :]

x_test = []
y_test = stock_data['Close'][training_data_len:].values
for i in range(60,len(test_data)):
    x_test.append(test_data[i-60:i,0])
    
x_test = np.array(x_test)

x_test = np.reshape(x_test, (x_test.shape[0],x_test.shape[1],1))

# Get the model's predicted price values
predictions = model.predict(x_test)
predictions = scaler.inverse_transform(predictions)

# Plot the test predictions
train = stock_data[:training_data_len]
valid = stock_data[training_data_len:]
valid['Predictions'] = predictions

plt.figure(figsize=(16,8))
plt.title(f'{ticker_symbol} Stock Price Prediction')
plt.xlabel('Date')
plt.ylabel('Price (USD)')
plt.plot(train['Close'])
plt.plot(valid[['Close', 'Predictions']])
plt.legend(['Train', 'Val', 'Predictions'], loc='lower right')
plt.show()
